src/common/utxobased/keymanager/cleaners.ts

```diff
@@ -0,0 +1,164 @@
+      }
+    },
+    clean => {
+      const { coinType, format, seed } = clean
+      return {
+        coinType,
+        format,
+        [`${coinName}Key`]: seed
+      }
+    }
+  )
+}
+
+/**
+ * A cleaner for the public key format following the key-formats specification.
+ *
+ * (spec: https://github.com/EdgeApp/edge-core-js/blob/master/docs/key-formats.md)
+ */
+export interface PublicKey {
+  publicKeys: {
+    [format in CurrencyFormat]?: string
+  }
+}
+export const asPublicKey: Cleaner<PublicKey> = asObject({
+  publicKeys: asObject({
+    bip32: asOptional(asString),
+    bip44: asOptional(asString),
+    bip49: asOptional(asString),
+    bip84: asOptional(asString)
+  })
+})
+
+/**
+ * This cleaner returns the wallet's supported formats as specified in the
+ * key-formats specification.
+ */
+export const asWalletKeysFormats: Cleaner<CurrencyFormat[]> = (
+  keys: any
+): CurrencyFormat[] => {
+  const format = asOptionalCurrencyFormat(keys?.format)
+
+  switch (format) {
+    case 'bip32':
+      return ['bip32']
+    case 'bip44':
+      return ['bip44']
+    case 'bip49':
+      return ['bip49', 'bip84']
+    case 'bip84':
+      return ['bip49', 'bip84']
+    default:
+      throw new Error(`Unsupported format ${format}`)
+  }
+}
+
+/**
+ * A cleaner that desensitizes the walletInfo object, excluding sensitive
+ * keys (seed/mnemonic, sync key, data key, etc). By using this object type
+ * internally within the plugin, we can minimize risk.
+ *
+ * It also includes internal derived data (publicKey, format, supportedFormats, etc).
+ * This derived data is to be used internally within the plugin and saved to disk (publicKey).
+ */
+export interface NumbWalletInfo {
+  id: string
+  type: string
+  keys: {
+    coinType: number
+    format: CurrencyFormat
+    supportedFormats: CurrencyFormat[]
+    publicKey: PublicKey
+  }
+}
+export const asNumbWalletInfo = (
+  engineInfo: EngineInfo
+): Cleaner<NumbWalletInfo> => {
+  return (walletInfo: any) => {
+    const { id, type, keys } = walletInfo
+    const { network: coinName, coinType, networkType } = engineInfo
+    const asCoinType = asOptional(asNumber, engineInfo.coinType)
+
+    // Public key may be missing from wallets pre-dating derivePublicKey implementation
+    // Get the public key from walletInfo or derive it from the private key
+    const publicKey = (_ => {
+      const publicKey = asMaybe(asPublicKey)(keys)
+      if (publicKey != null) {
+        return publicKey
+      }
+
+      const privateKey = asMaybe(asPrivateKey(coinName, coinType))(keys)
+      // Assertion error: we should at least have the private key
+      if (privateKey == null) {
+        throw new Error(
+          'Cannot derive public key from watch-only wallet (no private key)'
+        )
+      }
+
+      const currencyFormatKeys = deriveXpubsFromKeys({
+        privateKey,
+        coin: coinName,
+        network: networkType ?? NetworkEnum.Mainnet
+      })
+      return asPublicKey({
```

Comment:
This data has a known data type, so it should not need to be cleaned. Please replace the `asPublicKey` with a type delaration on the IIFE return value, so Typescript checks it instead of the cleaner.